Add monadic zygomorphism

kanterov opened the pull request:
I'm developing compiler with matryoshka, my original motivation is to annotate tree with types, and then compile, both phases operate under `M = Error \/ ?`. Basically, I want to do zygomorphism where both algebras are effectful.

I'm very new to recursion schemes, but I have a feeling there is something wrong because I wasn't able to describe `zygoM` in terms of `gcataM`. I'm not sure that this function is called `zygoM`, because probably it should be:

```scala
  def zygoM[A, B, M[_]: Monad]
    (t: T)
    (f: Algebra[Base, B], g: GAlgebraM[(B, ?), M, Base, A])
    (implicit BF: Traverse[Base])
      : M[A]
```

But then I'm not sure what is the right function for the particular use case I had.

I didn't add any tests because I'm not yet sure this is the right thing. I would be glad to get any help or feedback on this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slamdata/matryoshka?pullRequest=68) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slamdata/matryoshka?pullRequest=68) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slamdata/matryoshka?pullRequest=68) it.</sub>

kanterov commented:
Just realized that I my particular case I need something like

```scala
  def elgotZygoM[A, B, F[_]: Traverse, M[_]: Monad]
    (t: Fix[F])
    (f: AlgebraM[M, F, B], g: ElgotAlgebraM[(B, ?), M, F, A])
      : M[A]
```

sellout commented:
@kanterov Feel free to add that, too 😄

kanterov commented:
I added `zygoElgotM`, didn't touch tests yet, but it typechecked :)

kanterov commented:
@sellout I added tests. I had to extract zygomorphism tests because compilation time of `specs.scala` is terrible. Can you please take a look?

sellout commented on core/shared/src/main/scala/matryoshka/Recursive.scala:
```diff
@@ -101,13 +115,37 @@ trait Recursive[T] extends Based[T] {
+
+    def loop[W[_]: Comonad : Traverse]
+      (t: T)
+      (g: GAlgebraM[W, M, Base, A], k: DistributiveLaw[Base, (M ∘ W)#λ])
```
You can get rid of these parameters, because `loop` always uses the same `g` and `k`.

sellout commented on core/shared/src/main/scala/matryoshka/Recursive.scala:
```diff
@@ -101,13 +115,37 @@ trait Recursive[T] extends Based[T] {
+    (f: AlgebraM[M, Base, B], g: ElgotAlgebraM[(B, ?), M, Base, A])
+    (implicit BT: Traverse[Base])
+      : M[A] =
+    elgotCataM[A, B, (B, ?), M](t)(distZygoM(f, distApplicative[Base, M]), g)
```
Yeah, it’s interesting that you can use `distZygoM` here in `elgotCataM`, but not in `gCataM` above.

sellout commented on core/shared/src/main/scala/matryoshka/package.scala:
```diff
@@ -402,6 +402,18 @@ package object matryoshka {
     *
     * @group dist
     */
+  def distZygoM[F[_]: Functor, M[_]: Monad, B](
```
Nice.

kanterov commented:
@sellout I've updated `zygoM`, I had problem with implicit resolution, so I had to introduce `W[_]` before, I've cleaned it up and extract signature of `gcataM` as it probably should be. Could you please take a look?

kanterov commented:
Signing CLA is WIP, need to check with my employer first, but it shouldn't be a problem.

sellout commented on core/shared/src/main/scala/matryoshka/Recursive.scala:
```diff
@@ -101,13 +115,42 @@ trait Recursive[T] extends Based[T] {
+      : M[A] = {
+
+    // NB: Probably this is how `gcataM` should look like
+    def gcataM[W[_]: Comonad : Traverse]
```
I think this looks reasonable. How about you replace the existing `gcataM` and see if everything still works with this new def?

kanterov commented:
@sellout squashed, signed CLA, generalized `gcataM`, can you please take a look again? :)

sellout commented:
Trying to trigger a Travis build …

sellout commented:
I know this _looks_ mergable, but our merge script says it’s not. Can you merge master into this branch? Then we should be good.

kanterov commented:
@sellout I've rebased latest master on this branch.

Are these scripts public? Your release process looks awesome. 

By the way, I'm not familiar with it, but probably this change needs to be documented as breaking.

sellout commented:
Breaking because of the change to `gcataM`? This is what I get for not having better test coverage … I had _assumed_ that there were already uses of `gcataM` in definitions of other monadic, but alas that’s not true.

So, the gist of the breaking change is that `def cataM(t: T)(f: AlgebraM) = gcataM(t)(distCata, f)` will no longer work because we need to do something about the new `M ∘` in the `gcataM` `DistributiveLaw`, right?

sellout commented:
Also, our merge script is at https://github.com/slamdata/devtools/blob/master/bin/merge, and any of the publishing releases stuff is in https://github.com/slamdata/sbt-slamdata it’s not perfect, but I’m putting together a blog post that describes it and what parts of it we do & don’t like currently.

kanterov commented:
Yes, true, but I didn't find any open source codebase using `gcataM`, and it's possible to write a function from `DistributiveLaw[F, W]` to `DistributiveLaw[F, M ∘ W]` for any `M[_]: Applicative`, so I see it mostly as a change breaking binary compatibility.
